axlearn/common/optimizers.py

```diff
@@ -139,19 +140,40 @@ def update_fn(
     return PartitionedGradientTransformation(init=init_fn, update=update_fn, partition=partition_fn)
 
 
-def copy_partition(param_specs: NestedParameterSpec) -> NestedPartitionSpec:
+def copy_partition(
+    param_specs: Nested[ParameterSpec],
+    *,
+    pattern: Union[None, str, re.Pattern] = None,
+    memory_kind: Optional[MemoryKind] = None,
+) -> Nested[OptStateSpec]:
```

Comment:
Instead of coupling creation of `OptStateSpec` and setting of `memory_kind`, how about having a separate function for setting memory kind?

```
def set_memory_kind(opt_state_spec: Nested[OptStateSpec], *, pattern, memory_kind):
```

This allows `set_memory_kind` to be called multiple times, maybe for different memory kind. WDYT?

Reply:
I do not see how `set_memory_kind` will be different from `copy_partition`. Signature and implementation will be the same.

Reply:
Imagine in the future we have many types of memory kinds, e.g., "remote_host". Then we can do:

```
opt_state_specs = copy_partition(...)
opt_state_specs = set_memory_kind(..., "pinned_host")
opt_state_specs = set_memory_kind(..., "remote_host")
```

Reply:
Wouldn't it be the same as
```
opt_state_specs = copy_partition(...)
opt_state_specs = copy_partition(..., "pinned_host")
opt_state_specs = copy_partition(..., "remote_host")
```
Do you suggest using a separate function for slightly better readability?

Reply:
The difference is that `copy_partition` also performs the type conversion from `Nested[ParameterSpec]` to `Nested[OptStateSpec]`.

Reply:
I see. I can change the type of param_specs in `copy_partition` to `Nested[OptStateSpec]` since `ParameterSpec ` is a subclass of `OptStateSpec` and `copy_partition` doesn't use any new fields from `ParameterSpec`. Does this sound good?

Reply:
Thanks. SG.

Reply:
Done.